src/solvers/refinement/string_refinement.cpp

```diff
@@ -0,0 +1,1108 @@
+        symbol_exprt sym=to_symbol_expr(expr.lhs());
+        generator.set_string_symbol_equal_to_expr(sym, uncast);
+        return false;
+      }
+    }
+
+    if(refined_string_typet::is_unrefined_string_type(type))
+    {
+      symbol_exprt sym=to_symbol_expr(expr.lhs());
+      generator.set_string_symbol_equal_to_expr(sym, expr.rhs());
+      return false;
+    }
+    else if(type==ns.follow(expr.rhs().type()))
+    {
+      if(is_unbounded_array(type))
+        return true;
+      const bvt &bv1=convert_bv(expr.rhs());
+      const irep_idt &identifier=
+        to_symbol_expr(expr.lhs()).get_identifier();
+      map.set_literals(identifier, type, bv1);
+      if(freeze_all)
+        set_frozen(bv1);
+      return false;
+    }
+  }
+
+  return true;
+}
+
+/*******************************************************************\
+
+Function: string_refinementt::dec_solve
+
+ Outputs: result of the decision procedure
+
+ Purpose: use a refinement loop to instantiate universal axioms, 
+          call the sat solver, and instantiate more indexes if needed.
+
+\*******************************************************************/
+
+decision_proceduret::resultt string_refinementt::dec_solve()
+{
+  for(const exprt &axiom : generator.axioms)
+    if(axiom.id()==ID_string_constraint)
+    {
+      string_constraintt c=to_string_constraint(axiom);
+      universal_axioms.push_back(c);
+    }
+    else if(axiom.id()==ID_string_not_contains_constraint)
+    {
+      string_not_contains_constraintt nc_axiom=
+        to_string_not_contains_constraint(axiom);
+      array_typet witness_type
+        (refined_string_typet::index_type(),
+         infinity_exprt(refined_string_typet::index_type()));
+      generator.witness[nc_axiom]=
+        string_exprt::fresh_symbol("not_contains_witness", witness_type);
+      not_contains_axioms.push_back(nc_axiom);
+    }
+    else
+    {
+      add_lemma(axiom);
+    }
+
+  initial_index_set(universal_axioms);
+  update_index_set(cur);
+  cur.clear();
+  add_instantiations();
+
+  while((initial_loop_bound--)>0)
+  {
+    decision_proceduret::resultt res=supert::dec_solve();
+
+    switch(res)
+    {
+    case D_SATISFIABLE:
+      if(!check_axioms())
+      {
+        debug() << "check_SAT: got SAT but the model is not correct" << eom;
+      }
+      else
+      {
+        debug() << "check_SAT: the model is correct" << eom;
+        return D_SATISFIABLE;
+      }
+
+      debug() <<  "refining..." << eom;
+      // Since the model is not correct although we got SAT, we need to refine
+      // the property we are checking by adding more indices to the index set,
+      // and instantiating universal formulas with this indices.
+      // We will then relaunch the solver with these added lemmas.
+      current_index_set.clear();
+      update_index_set(cur);
+      cur.clear();
+      add_instantiations();
+
+      if(variable_with_multiple_occurrence_in_index)
```

Comment:
Actually I put this warning to signal that the algorithm is not guaranteed to terminate when multiple occurrence occurs, but since we now put a bound on the number of loops, we don't need to keep it.
Do you prefer to have it removed?